### ISSUE TITLE
Image widget alt and title attributes

### ISSUE THREAD

**Issue by afercia** [xwp/wp-core-media-widgets#47]
Just quickly tested the Image widget:
- seems the `alt` attribute is not set on the preview in the WIdgets screen or in the Customizer sidebar
- in the frontend instead, it is set. But there's also a `title` attribute that shouldn't be there.

Also, I'm not sure about _which_ alt attribute is used. The one from the media library or the one from the "Image Details" media modal"? Which one should prevail? And should one of those actually overwrite the other one or maybe users want to keep one default alt attribute in the Media Library and have a different alt attribute in the frontend, depending on the actual use context of the image?

**Comment by afercia** [comment]
Also, when the image has no alt text value and the user doesn't enter an alt, that's perfectly ok in the front-end. An empty `alt=""` indicates the image is purely decorative and that's the assumed intention of the author. Screen readers will ignore the image on the front-end, and that's ok.

However, also the previews in the admin will have an empty `alt=""`. This way, screen reader users won't have any feedback about the image they've selected and saved in the widget. This is an edge case: when there's no `alt` value, either the alt attribute should be filled in with _something_ or the alt attribute should be completely removed. Clearly, in an authoring context there's the need to inform users about the selected image. In the front-end instead, the alt can be empty.

The first option (filling the alt with something) would be preferable. The second option, while counterintuitive at first sight, would allow screen readers to read out the file name (they look for the src attribute):

![screen shot 2017-04-05 at 22 50 09](https://cloud.githubusercontent.com/assets/1682452/24726678/817e211a-1a53-11e7-83ac-1e9e18589b00.png)

**Comment by westonruter** [comment]
@afercia:

> seems the alt attribute is not set on the preview in the Widgets screen or in the Customizer sidebar

The template was adapted from the Media control which likewise does not populate the `alt` attribute: https://github.com/WordPress/wordpress-develop/blob/4.7.3/src/wp-includes/customize/class-wp-customize-media-control.php#L162

I'm not saying this is correct, but it is just what is already in core.

> in the frontend instead, it is set. But there's also a title attribute that shouldn't be there.

The image details modal in core allows for the title attribute to be set, so that is where the `title` is coming from:

<img src="https://cloud.githubusercontent.com/assets/134745/24731637/3490ad7e-1a20-11e7-83ed-c6109e0591c2.png" width=300>
 
> Also, I'm not sure about which alt attribute is used. The one from the media library or the one from the "Image Details" media modal"? Which one should prevail? And should one of those actually overwrite the other one or maybe users want to keep one default alt attribute in the Media Library and have a different alt attribute in the frontend, depending on the actual use context of the image?

The initial `alt` should come from the `attachment` saved in the DB. Thereafter, it can be modified via the Image Details model to be contextual to the widget instance.

**Comment by westonruter** [comment]
@afercia could you please review #50 and confirm whether it addresses your concerns here?

**Comment by rianrietveld** [comment]
Concerning the title attribute: 
Screen readers behave differently when announcing alt and title attributes. That's why it's a bit tricky to add the title attribute. And that's why we removed it from links in core.

But concerning images I think there is a difference. The title attribute is not shown on touch devices, so people probably won't put crucial info in it and apparently it has [SEO value ](https://yoast.com/image-seo/). Testing with Voiceover, only the alt attribute is announced.
Not sure if we need to be as strict here as we are on links? Check on more AT maybe?
I installed the widget on our test server: http://wpaccess.org/trunk/src/blog/

About the Image Details:
I agree with Andrea, having an empty alt attribute here is confusing, the image is not announced.
Leaving the alt attribute out completely in the Image Details will result announcing the URL and that is maybe exactly what you want.
